### DISCUSSION TITLE
详解 DNS 与 CoreDNS 的实现原理 · /dns-coredns

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#533] [Announcements]
https://draveness.me/dns-coredns

**Comment by x1angli** [comment]
赞！大咖的这篇文章已经被一些第三方内容聚合推荐平台给收录了

关于DNS解析，貌似从桌面端的DNS客户端的解析结果会类似薛定谔的猫一样不可预测。比如，在调用`nslookup`命令时，即使设置了`set norecurse`参数，查询结果还是会包含non-authorative answer。

另外，期待大咖有空可以讲解一下coreDNS与服务注册

**Comment by winneryong** [comment]
请问这图是用什么软件画的？

**Comment by winneryong** [comment]
看到了其他文章回答了这个问题，谢谢，图非常赞

**Comment by draveness** [comment]
> @winneryong
> 看到了其他文章回答了这个问题，谢谢，图非常赞

每个下面我都写了。。

**Comment by ybintian** [comment]
能不能分享下博主的画关系图的工具 感觉非常不错。

**Comment by winneryong** [comment]
@ybintian Sketch

**Comment by ybintian** [comment]
> @winneryong
> @ybintian Sketch

谢谢了

**Comment by sunencheng** [comment]
大佬，未登录状态下，点击menu，选server，会报重定向次数太多，无法访问的错误。

**Comment by draveness** [comment]
> @sunencheng
> 大佬，未登录状态下，点击menu，选server，会报重定向次数太多，无法访问的错误。

好的，我看看

**Comment by draveness** [comment]
> @sunencheng
> 大佬，未登录状态下，点击menu，选server，会报重定向次数太多，无法访问的错误。

已经都删掉了，哈哈哈

**Comment by sunencheng** [comment]
> @draveness
> > @sunencheng
> > 大佬，未登录状态下，点击menu，选server，会报重定向次数太多，无法访问的错误。
> 
> 已经都删掉了，哈哈哈

嘿嘿，我是不是要清空缓存啥的来解决这个问题啊？

**Comment by sunencheng** [comment]
> @draveness
> > @sunencheng
> > 大佬，未登录状态下，点击menu，选server，会报重定向次数太多，无法访问的错误。
> 
> 已经都删掉了，哈哈哈

因为我还在报错呀。

**Comment by sunencheng** [comment]
> @draveness
> > @sunencheng
> > 大佬，未登录状态下，点击menu，选server，会报重定向次数太多，无法访问的错误。
> 
> 已经都删掉了，哈哈哈

原来是把菜单给删除了😂

**Comment by llussy** [comment]
大佬您好，有个问题请教下    我的k8s集群使用的coredns，发现一个问题 。 在pod中dig一个域名，有时解析到内网有时解析到外网。
我查看我的coredns配置
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns
  namespace: kube-system
data:
  Corefile: |
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          upstream
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        proxy . /etc/resolv.conf
        cache 60
        reload
        loadbalance
    }
```
之前怀疑是loadbalance这个参数导致的，去掉后测试也不行。

我的coredns所在宿主机的/etc/resolv.conf中既有内网dns也有114dns。
大佬 这个需要怎么操作 才能让pod只解析到内网？

**Comment by dawncold** [comment]
最终"队"会调用

---


*2020-05-14 UPDATES：已修复*

**Comment by langcui** [comment]
请问 13个根域名服务器的ip是hardcode吗?

**Comment by draveness** [comment]
> 请问 13个根域名服务器的ip是hardcode吗?

可以了解一下根域名服务器的域名和对应的 IP 地址 https://www.iana.org/domains/root/servers

自己也可以跑一下 dig 看看

```bash
$ dig a.root-servers.net
```

**Comment by ZhouFyk** [comment]
「CoreDNS 可以通过四种方式对外直接提供 DNS 服务，分别是 UDP、gRPC、HTTPS 和 TLS：
![coredns servers](https://img.draveness.me/2018-11-07-coredns-servers.png)
」
这 4 种「方式」中 UDP / TLS / HTTPS 作为网络协议在不同的层次上，而 gRPC 则是一个框架，他们不是同一类型同一层次的，这么分类让我理解起来有点迷糊，或者是否是我的理解有误？

**Comment by draveness** [comment]
> 「CoreDNS 可以通过四种方式对外直接提供 DNS 服务，分别是 UDP、gRPC、HTTPS 和 TLS： 」
> 这 4 种「方式」中 UDP / TLS / HTTPS 作为网络协议在不同的层次上，而 gRPC 则是一个框架，他们不是同一类型同一层次的，这么分类让我理解起来有点迷糊，或者是否是我的理解有误？

这里跟协议层级没关系，不要从这个角度理解，只是说可以通过哪些渠道

**Comment by npc64** [comment]
"DNS 解析器首先会向就近的根 DNS 服务器 . 请求顶级域名 DNS 服务的地址" 这个说法也不是很准确，应该说，解析器会向根查询 draveness.me.

**Comment by npc64** [comment]
> @langcui
> 请问 13个根域名服务器的ip是hardcode吗?


是 hardcode 的，递归解析器里通常会 hardcode 一份，当然也支持配置，~~但应该不会动态更新。~~
更正一下，bind9 看来是会动态更新，https://kb.isc.org/docs/aa-01309

**Comment by draveness** [comment]
>  @laziguy "DNS 解析器首先会向就近的根 DNS 服务器 . 请求顶级域名 DNS 服务的地址" 这个说法也不是很准确，应该说，解析器会向根查询 draveness.me.

因为这篇文章也比较久了，如果有问题的话，建议给出相关的引用以及为什么这里会有问题，重新找回上下文是一件很困难的事情，谢谢
